### DISCUSSION TITLE
如何优雅地退出后台守护运行的nonebot v1程序

### DISCUSSION THREAD

**Discussion by CJReinforce** [nonebot/discussions#157] [问答] [answered]
如果是前台运行，我知道按下ctrl+c即可，但或许有个问题就是：程序退出前我没有代码去关闭与数据库的连接，或许会造成不好的影响。而大部分情况下，我是用nohup后台守护运行nonebot程序，这个时候我只会kill进程来关闭了，但同样没有办法执行关闭数据库连接。我还想着用代码来实现关闭程序，监听一个命令，接收到命令后就先关闭数据库连接，然后sys.exit()或者raise keyboardinterrupt。我不知道自己的做法是否正确，或者有更优雅的方式？

**Comment by yanyongyu** [comment, marked as answer]
或许可以使用以下代码，在关闭时处理数据库连接

```python
@get_bot().server_app.after_serving
async def shutdown():
    ...
```

**Reply by CJReinforce** [reply]
这样或许可以在前台运行并按下ctrl+c的时候执行shutdown函数，那在后台kill进程的时候会不会没机会执行这个函数呢？以及感觉kill进程还有点麻烦，得连服务器查进程号再kill，我更想能直接通过代码来远程关闭它（比如机器人监听了一个命令`stop`，接收到命令后就执行了一段代码自动关了机器人）。我折腾了一下午，目前自己的做法是这样的：

建一个插件，代码如下：
```python
import nonebot
from db_init import cur, conn
import asyncio


async def cancel_task(task):
    task.cancel()
    try:
        await asyncio.sleep(10)
    except asyncio.CancelledError:
        nonebot.log.logger.info('Task cancelled.')


@nonebot.on_command('stop', aliases=('停止', '结束'), permission=lambda sender: sender.is_superuser)
async def _(session: nonebot.CommandSession):
    # 关闭数据库连接
    cur.close()
    conn.cancel()
    conn.close()
    # 消息提示
    await session.send('机器人即将关闭')
    # 取消每个任务
    tasks = [cancel_task(task) for task in asyncio.all_tasks()]
    nonebot.log.logger.info('QQ robot exiting... Cancell all {} tasks.'.format(len(tasks)))
    await asyncio.wait(tasks)
    # 当任务关闭完成，关闭事件循环
    nonebot.log.logger.info('Stopping event loop...')
    asyncio.get_event_loop().stop()
    nonebot.log.logger.info('Event loop stopped.')
```
`cancel_task`函数之所以睡十秒并捕捉异常是因为看到[cancel()官方文档的声明](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.cancel)
>"This arranges for a CancelledError to be thrown into the wrapped coroutine on **the next cycle** through the event loop."

所以硬睡十秒等下一个周期来到
然后bot.py修改如下：
```python
import nonebot
import config
from os import path
from asyncio import CancelledError


if __name__ == '__main__':
    nonebot.init(config, start_scheduler=True)
    nonebot.load_builtin_plugins()
    nonebot.load_plugins(path.join(path.dirname(__file__), 'qq', 'plugins'), 'qq.plugins')
    try:
        nonebot.run(host='127.0.0.1', port=54321)
    except CancelledError:
        pass
    nonebot.log.logger.info('QQ robot exits.')
```
当我给机器人发命令`stop`后，日志输出如下：
```python
[2022-05-29 17:21:07,060 nonebot] INFO: Self: 机器人qq号, Message -564871690 from 我的qq号: 'stop'
[2022-05-29 17:21:07,064 nonebot] INFO: QQ robot exiting... Cancell all 9 tasks.
[2022-05-29 17:21:07,066 nonebot] INFO: Task cancelled.
[2022-05-29 17:21:07,066 nonebot] INFO: Task cancelled.
[2022-05-29 17:21:07,066 nonebot] INFO: Task cancelled.
[2022-05-29 17:21:07,066 nonebot] INFO: Task cancelled.
[2022-05-29 17:21:07,066 nonebot] INFO: Task cancelled.
[2022-05-29 17:21:07,066 nonebot] INFO: Task cancelled.
[2022-05-29 17:21:07,066 nonebot] INFO: Task cancelled.
[2022-05-29 17:21:07,066 nonebot] INFO: Task cancelled.
[2022-05-29 17:21:07,066 nonebot] INFO: Task cancelled.
[2022-05-29 17:21:07,066 nonebot] INFO: QQ robot exits.
```
当然，我qq也收到了消息提示`机器人即将关闭`
这样看起来好像是数据库连接也关闭了，所有机器人相关的任务也成功取消了，然后安全退出了。就是关闭事件循环的这几句没运行：
```python
nonebot.log.logger.info('Stopping event loop...')
asyncio.get_event_loop().stop()
nonebot.log.logger.info('Event loop stopped.')
```
不知道以上是不是一次把所有善后工作处理到位然后安全退出的行为。由于对异步框架知之甚少，还望大佬不吝赐教。

**Reply by mnixry** [reply]
事实上，在类Unix操作系统中，你输<kbd>Ctrl</kbd>+<kbd>C</kbd>，在实际上是向程序发送了`SIGINT`信号。所以，如果你想要优雅地终止一个运行在后台的进程，只需要向它发送`SIGINT`信号，即可达到和你按<kbd>Ctrl</kbd>+<kbd>C</kbd>一样的效果。（大多数进程守护工具就是这样做的！）

P.S. 在Linux下强制终止进程是`SIGKILL`信号，既众所周知的`kill -9`

P.P.S. 关于更多的操作系统信号，你可以参考Python的[signal标准库](https://docs.python.org/zh-cn/3/library/signal.html)文档

P.P.P.S. 关于如何在异步程序中处理进程终止信号，你可以参考Python文档的[为SIGINT和SIGTERM设置信号处理器](https://docs.python.org/zh-cn/3/library/asyncio-eventloop.html#set-signal-handlers-for-sigint-and-sigterm)一节

**Reply by CJReinforce** [reply]
@mnixry 非常感谢！您提供的资料让我获益匪浅！
目前写了这段代码用来在关闭时处理数据库连接：
```python
from db_init import cur, conn
import nonebot
import os
from signal import SIGINT


bot = nonebot.get_bot()


@bot.server_app.after_serving
async def shutdown():
    # 关闭数据库连接
    cur.close()
    conn.cancel()
    conn.close()
    # 输出日志
    nonebot.log.logger.info('QQ robot exiting...')
    nonebot.log.logger.info('Connection to database closed.')
```
P.S. 注意到直接写`@nonebot.get_bot().server_app.after_serving`的话，会报语法错误

这样前台运行的时候按<kbd>Ctrl</kbd>+<kbd>C</kbd>就可以执行`shutdown`函数的内容了。进程守护工具我换成了pm2，注意设置`--kill-timeout`参数（我设了5000），不然默认的1600ms经常程序还没来得及响应pm2就以为关不掉它于是就开始发`SIGKILL`信号了。然后因为执念，我还是写了一个监听命令来远程关掉它：
```python
@nonebot.on_command('stop', aliases=('停止', '结束'), permission=lambda sender: sender.is_superuser)
async def _(session: nonebot.CommandSession):
    pid = os.getpid()
    # 消息提示
    await session.send('机器人即将关闭')
    os.kill(pid, SIGINT)
```
这样发送命令`stop`也能关掉它了，但是这个方式要注意设置pm2的`--no-autorestart`参数，否则关了之后pm2还会再自动启动它。

综上，在我常用的几种使用方式里，已经能够优雅地退出了。再次表示感谢！

**Reply by yanyongyu** [reply]
PS，直接写`@get_bot().server_app.after_serving`会报错是因为你的Python版本低了

**Comment by 2catycm** [comment]
请问nonebot2中如何进行类似的操作？

**Reply by 2catycm** [reply]
nonebot2是否有收到SIGINT\SIGTERM\SIGKILL退出会调用的函数

**Reply by yanyongyu** [reply]
你可以参考[文档](https://nonebot.dev/docs/next/advanced/runtime-hook#%E7%BB%88%E6%AD%A2%E5%A4%84%E7%90%86)
